File: config_rules_poc.py
import boto3
import os
import json
from bson import json_util
AWS_DEFAULT_REGION = 'us-west-1'
AWS_ACCESS_KEY_ID = 'access-key-id'
AWS_SECRET_ACCESS_KEY = 'secret-access-key'

def set_aws_config():
    # do the credential setting here
    #os.environ['AWS_ACCESS_KEY_ID'] = AWS_ACCESS_KEY_ID
    #os.environ['AWS_SECRET_ACCESS_KEY'] = AWS_SECRET_ACCESS_KEY
    pass

def get_config_client(region):
    set_aws_config()
    session = boto3.session.Session()
    client = session.client('config', region_name=region)
    return client

def get_config_rules(client):
    next_token = ""
    config_rules = []
    while True:
        response = client.describe_config_rules()
        if isRequestSuccessful(response):
            if response['ConfigRules']:
                config_rules.extend(response['ConfigRules'])
        next_token = response.get("NextToken")
        if not next_token:
            break
    return config_rules

def get_compliance_details(client, rule_name):
    next_token = ""
    compliance_details = []
    while True:
        response = client \
            .get_compliance_details_by_config_rule(ConfigRuleName=rule_name,
                                                   NextToken=next_token)
        if isRequestSuccessful(response):
            if response["EvaluationResults"]:
                compliance_details.extend(response["EvaluationResults"])
        next_token = response.get("NextToken")
        if not next_token:
            break
    return compliance_details

def get_compliance_summary(client):
    response = client \
        .get_compliance_summary_by_config_rule()
    if isRequestSuccessful(response):
        #print(response)
        #print(json.dumps(response, default=json_util.default))
        return response["ComplianceSummary"]

def isRequestSuccessful(response):
    return response["ResponseMetadata"]["HTTPStatusCode"] == 200

def main():
    # get config rules
    client = None
    regions = [AWS_DEFAULT_REGION]

    config_rules = []
    config_rules_details = []

    for region in regions:
        client = get_config_client(region)
        if client:
            config_rules = get_config_rules(client)
            # check compliance of each rule
            for config_rule in config_rules:
                config_rule_details = get_compliance_details(client, config_rule["ConfigRuleName"])
                config_rules_details.extend(config_rule_details)
                #config_rule_summary = get_compliance_summary(client)
                #config_rules_summary.append(config_rule_summary)

    print(json.dumps(config_rules, default=json_util.default))
    print(json.dumps(config_rules_details, default=json_util.default))



if __name__=='__main__':
    main()

File: config_rules_poc.go
package main

import (
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/configservice"
	"github.com/aws/aws-sdk-go/service/configservice/configserviceiface"
	"github.com/aws/aws-sdk-go/aws"
	"fmt"
)

const (
	AwsDefaultRegion = "us-west-1"
	AwsAccessKeyId = "access-key-id"
	AwsSecretAccessKey = "secret-access-key"
)

func setAwsConfig() {
	// do the credential setting here
	//os.Setenv("AWS_ACCESS_KEY_ID", AwsAccessKeyId)
	//os.Setenv("AWS_SECRET_ACCESS_KEY", AwsSecretAccessKey)
}

func getConfigClient(region string) configserviceiface.ConfigServiceAPI {
	sess := session.Must(session.NewSession())
	client := configservice.New(sess, aws.NewConfig().WithRegion(region))
	return client
}

func getConfigRules(client configserviceiface.ConfigServiceAPI) (
	[]*configservice.ConfigRule, error) {
	output, err := client.DescribeConfigRules(&configservice.DescribeConfigRulesInput{
	})
	return output.ConfigRules, err
}

func getComplianceDetails(client configserviceiface.ConfigServiceAPI, rule_name string) (
	[]*configservice.EvaluationResult, error) {
	output, err := client.GetComplianceDetailsByConfigRule(
		&configservice.GetComplianceDetailsByConfigRuleInput{
			ConfigRuleName: &rule_name,
		})
	return output.EvaluationResults, err
}

func main() {
	setAwsConfig()
	regions := [1]string{AwsDefaultRegion}
	configRulesResult := make([]*configservice.ConfigRule, 0)
	configRulesDetailsResult := make([]*configservice.EvaluationResult, 0)
	for _, region := range regions {
		client := getConfigClient(region)
		configRules, err := getConfigRules(client)
		configRulesResult = append(configRulesResult, configRules...)
		if err != nil {
			fmt.Println(err)
			continue
		}
		for _,configRule := range configRules {
			evaluationResults, err := getComplianceDetails(client, *configRule.ConfigRuleName)
			if err != nil {
				fmt.Println(err)
				continue
			}
			configRulesDetailsResult = append(configRulesDetailsResult, evaluationResults...)
		}
	}
	fmt.Println(configRulesResult)
	fmt.Println(configRulesDetailsResult)
}